### DISCUSSION TITLE
Adjust the positioning of PDF content

### DISCUSSION THREAD

**Discussion by JackieFei** [pymupdf/PyMuPDF#942] [Looking for help] [answered]
Is there any way to adjust the positioning of PDF content?

- without converting to image/svg format for text editing later.
- without rewriting text to avoid missing any attribute.

For example:
```
PDF has text "Hello World" at (top, left), then want to move this text to (top+10, left+10).
```

Thanks in advanced.

**Comment by JorjMcKie** [comment, marked as answer]
> Is there any way to adjust the positioning of PDF content?

You mean some specific text only, and not the whole page content, right?
That won't really work: you have to find the place, where that text piece is defined in the page `/Contents` and then modify its text matrix `Tm`.
Problems are manifold:
* In its **_simplest_ (!!!)** form, the text object may look like this:
```postscript
BT
/helv 11 Tf
1 0 0 1 100 742 Tm  % = point (100,100) on a A4 page
[(Hello Earth!)] TJ
ET
```

or like this (text encoded before `TJ` is "Hello Mars!" - undetectable in any practicable way):
```postscript
BT
/F0 11 Tf
1 0 0 1 100 692 Tm  % = point (100, 150)
[<00290046004D004D00500001002E0042005300540002>] TJ
ET
```

* then, the text could be multi-line - with or without its own `Tm`
* the `Tm` may not contain the starting position so clearly like here: previous matrices may exist (somewhere!), that scale or otherwise modify the page geometry. Those would have to be found and concatenated to our `Tm` delivering `Point(100, 100)` in the end
* then the text may not even be on the page's `/Contents`, but in some Form XObject ...

So, the short answer to your question is "no".

**Reply by JackieFei** [reply]
Thanks for the explaination about the specific text only.

Is it possible about the whole page contents moved from (0, 0) to (10, 10)?
> maybe put/merge a whole text page into another empty page at location (10, 10)

**Reply by JorjMcKie** [reply]
> Is it possible about the whole page contents moved from (0, 0) to (10, 10)?

Yes, that works using a clean approach:
1. make a new page in a new pdf
2. define the rectangle in which to show the original page: `rect = fitz.Rect(10, 10, newpage.rect.x1, newpage.rect.y1)`. Or, the same but more obvious: `rect = newpage.rect + (10, 10, 0, 0)`
3. define the rectangle of the source page, which we want to show: this starts from its (0,0) and ends 10 points less in each dimension `clip = fitz.Rect(0, 0, oldpage.rect.x1 - 10, oldpage.rect.y1 - 10)`. If we start the display shifted by displacement (10, 10), we must of course cut those values off of the rectangle's width and height. Again: `clip = oldpage.rect + (0, 0, -10, -10)` also works.
4. then `newpage.show_pdf_page(rect, oldpdf, oldpage.number, clip=clip)`.

Note: `rect` and `clip` do not have to have identical dimensions. The source is always scaled to fill the target rect - keeping aspect ratios by default (can be switched off, too).

**Reply by JackieFei** [reply]
<img width="873" alt="截圖 2021-03-11 下午6 12 16" src="https://user-images.githubusercontent.com/12691750/110771371-61f33980-8295-11eb-8019-1a9293f41519.png">

Does the `show_pdf_page` generate a vector image in newpage? Is it editable by pdf editor (text mode)?

**Reply by JorjMcKie** [reply]
> Does the show_pdf_page generate a vector image in newpage?

Yes!

> Is it editable by pdf editor (text mode)?

I don't know pdf editor.

**Reply by JorjMcKie** [reply]
Another, less clean alternative: Change the page's mediabox by low-level code.
```python
>>> page.xref
4
>>> doc.xref_get_key(4, "MediaBox")
('array', '[0 0 595 842]')
>>> doc.xref_set_key(4,"MediaBox", "[-10 10 585 852]")  # effectively add (-10, 10, -10, 10)
```

Example text search shows the shift:
```python
>>> doc1[0].search_for("hello world!")
[Rect(100.0, 88.17500305175781, 157.45298767089844, 103.28900146484375)]
>>> doc2[0].search_for("hello world!")
[Rect(110.0, 98.17500305175781, 167.45298767089844, 113.28900146484375)]
>>>
```
The top-left and bottom-right coordinates both are 10 points larger.
